iConverter/Services/Converter/ConverterService.swift

```diff
@@ -6,129 +6,61 @@
     
-    // MARK: - Inputs
-    let convert: PublishRelay<Transaction> = .init()
-    
-    // MARK: - Outputs
-    let baseState = PublishRelay<BaseState>()
-    let onSuccess = PublishRelay<Transaction>()
-    let onError = PublishRelay<String>()
-    
     init(
-        converterApi: ConverterApiProtocol,
+        convertAPIService: ConverterAPIServiceProtocol,
         balanceDataStore: BalanceDataStoreProtocol,
         converterValidator: ConverterValidatorProtocol,
         feeService: FeeServiceProtocol,
         historyService: HistoryDataStoreProtocol
     ) {
-        self.converterApi = converterApi
+        self.convertAPIService = convertAPIService
         self.balanceDataStore = balanceDataStore
         self.converterValidator = converterValidator
         self.feeService = feeService
         self.historyService = historyService
-        
-        doBindings()
     }
-}
 
-// MARK: - Do bindings
-extension ConverterService {
-    func doBindings() {
-        let checkRequest = convert
-            .withLatestFrom(balanceDataStore.allBalancies) { transaction, balancies -> (transaction: Transaction, balance: Balance)? in
-                let balance = balancies.first { $0.currency == transaction.fromCurrency }
-                guard let balance = balance else {
-                    return nil
-                }
-                return (transaction, balance)
-            }
-            .filterNil()
-            .map { [weak self] t -> Transaction? in
-                if let error = self?.validate(t.transaction, balance: t.balance) {
-                    self?.onError.accept(error)
-                    return nil
-                }
-                
-                return t.transaction
-            }
-            .filterNil()
-        
-        checkRequest
-            .bind(to: feeService.addFee)
-            .disposed(by: disposeBag)
-        
-        let afterFee = feeService.updatedTransaction
-                
-        let response = checkRequest
-            .flatMap { [converterApi] request in converterApi.convert(with: request) }
-            .share()
-        
-        let transaction = response.compactMap(\.response)
-            .withLatestFrom(afterFee) { $1.copy(converted: $0.amount.toDouble()) }
-        
-        transaction
-            .map { $0.copy(createdDate: .now) }
-            .bind(to: historyService.saveTransaction)
-            .disposed(by: disposeBag)
-        
-        transaction
-            .subscribe(onNext: { [balanceDataStore] t in
-                balanceDataStore.updateBalance.accept(t)
-                balanceDataStore.update.accept(())
-            }).disposed(by: disposeBag)
-        
-        transaction
-            .map { _ in () }
-            .bind(to: historyService.updateHistory)
-            .disposed(by: disposeBag)
+    func convert(transaction: Transaction) async throws -> Transaction {
+        let allBalance = balanceDataStore.loadAllBalances()
```

Comment:
I think for this one, it's better to expose `func getBalance(for currency: Currency) -> Balance` so you won't have to query all?